tokio/src/io/util/vec_with_initialized.rs

```diff
@@ -1,6 +1,19 @@
 use crate::io::ReadBuf;
 use std::mem::MaybeUninit;
 
+mod private {
+    pub trait Sealed {}
+
+    impl Sealed for Vec<u8> {}
+    impl Sealed for &mut Vec<u8> {}
+}
+
+/// A sealed trait that constrains the generic type parameter in `VecWithInitialized<V>`.  That struct's safety relies
+/// on certain invariants upheld by `Vec<u8>`.
+pub(crate) trait VecU8: AsMut<Vec<u8>> + private::Sealed {}
```

Comment:
The question is if it is better to just make the trait unsafe instead of sealing it.

Reply:
An unsafe trait could work as well.  I felt that a sealed safe trait would be a bit better in that:
* minimizes `unsafe`s
* sealing and defining a controlled set of impls seemed pretty succinct and to the point

However, I'm open to hearing arguments for making it an unsafe trait instead.  Either way, the thing that somewhat irked me with existing code is that we weren't capturing safety via types.

Reply:
I mean, either is fine. I was mostly wondering to myself if sealing even works without a crate boundary, but after some more thought, I guess it does.

Reply:
Yeah, it's essentially sealing from the rest of the crate.  This allows specifying the valid types in the `vec_with_initialized` module, which is nice since it's the same source file as the uses and where the commentary about `Vec`'s invariants lives.